File: src/main/java/com/qtech/bubbles/common/text/translation/I18n.java
package com.qtech.bubbles.common.text.translation;

import java.util.Locale;

public class I18n {
    private static final LanguageMap localizedName = LanguageMap.getInstance();
    private static final LanguageMap fallbackTranslator = new LanguageMap(new Locale("en", "us"));

    public static String translateToLocal(String key) {
        if (canTranslate(key)) {
            return localizedName.translateKey(key);
        } else {
            return translateToFallback(key);
        }
    }

    public static String translateToLocalFormatted(String key, Object... format) {
        if (canTranslate(key)) {
            return localizedName.translateKeyFormat(key, format);
        } else {
            return translateToFallbackFormatted(key, format);
        }
    }

    public static String translateToFallback(String key) {
        return fallbackTranslator.translateKey(key);
    }

    public static String translateToFallbackFormatted(String key, Object... format) {
        return fallbackTranslator.translateKeyFormat(key, format);
    }

    public static boolean canTranslate(String key) {
        return localizedName.isKeyTranslated(key);
    }

    public static long getLastTranslationUpdateTimeInMilliseconds() {
        return localizedName.getLastUpdateTimeInMilliseconds();
    }
}

File: src/main/kotlin/qtech/bubbles/common/text/translation/I18n.kt
package qtech.bubbles.common.text.translation

import java.util.*

object I18n {
    private val localizedName: LanguageMap = LanguageMap.Companion.instance
    private val fallbackTranslator = LanguageMap(Locale("en", "us"))
    @JvmStatic
    fun translateToLocal(key: String): String? {
        return if (canTranslate(key)) {
            localizedName.translateKey(key)
        } else {
            translateToFallback(key)
        }
    }

    @JvmStatic
    fun translateToLocalFormatted(key: String, vararg format: Any?): String? {
        return if (canTranslate(key)) {
            localizedName.translateKeyFormat(key, *format)
        } else {
            translateToFallbackFormatted(key, *format)
        }
    }

    fun translateToFallback(key: String): String {
        return fallbackTranslator.translateKey(key)
    }

    fun translateToFallbackFormatted(key: String, vararg format: Any?): String {
        return fallbackTranslator.translateKeyFormat(key, *format)
    }

    fun canTranslate(key: String?): Boolean {
        return localizedName.isKeyTranslated(key)
    }

    val lastTranslationUpdateTimeInMilliseconds: Long
        get() = localizedName.lastUpdateTimeInMilliseconds
}